### ISSUE TITLE
Typescript/Jest performance issues

### ISSUE THREAD

**Issue by tombanksme** [date-fns/date-fns#3806]
We've recently updated one of our typescript projects from date-fns `^2.30.0` to `^3.6.0`. Since then we've seen massive increases in runtime for our jest tests; on my M1 MacBook Pro our test suite increased from 87s to 292s after updating `date-fns`.

```
typescript == 5.4.5
ts-jest == 29.1.2
jest == 29.7.0
date-fns == 3.6.0
```

If I rollback using `npm install date-fns@2` the test duration returns to normal. Has anyone else ran into this issue?

**Comment by fturmel** [comment]
That's a little puzzling... If anything it should be a bit faster because there are less conversions and checks at runtime.

A focused reproduction of the problem would be helpful.

Some ideas:

* Are all tests passing and equally slower, or can you isolate specific tests that got slower or are timing out?
* Could you test with older v3 versions? Specifically v3.1.0 (in case it's related to #3769) and v3.5.0 (before CDN re-introduction which added more files to the npm package, maybe ts-jest is transpiling more than it should?)
* Did you review the v3 release notes and audit your app for breaking changes? https://github.com/date-fns/date-fns/releases https://github.com/date-fns/date-fns/releases/tag/v3.0.0

**Comment by tombanksme** [comment]
Thanks for the response. I've found the problem (reproduction repo). We use `@node20/tsconfig` as our base Typescript configuration which sets `moduleResolution` to `node16`

If you remove it from the `tsconfig.json` the runtime of the tests in the reproduction repo go from ~1.3s to ~0.5s. This change doesn't break anything in our project; I'm not sure how you want to document/fix this incompatability in date-fns.

**Comment by TheSamDickey** [comment]
Hello! I upgraded the packages for my service and scratched my head for a long time why the unit tests were taking forever to begin running. I boiled it down to date-fns, and my setup is nearly identical to @tombanksme. I noticed the same behavior on my M1 MacBook Pro as well as in the Azure DevOps pipelines. I don't think it's exclusive to my M1 MacBook Pro. 

I saw a difference in ~2 minutes on version 2.30.0 to ~8 minutes after upgrading to 4.1.0.

Please let me know if there are any other details I can provide.

Thank you!

**Comment by rryter** [comment]
Same problem here, the upgrade totally wrecked our test performance:

2.30.0 -> 10 minutes 
4.1.0. -> 45+ minutes

If I get the chance I'll look into setting up a reproduction env.

**Comment by mindplay-dk** [comment]
ugh, I've been pestering the Vitest team for a week, thinking they were doing something wrong. 🥴

**Comment by mindplay-dk** [comment]
I have a simple test script here:

https://stackblitz.com/edit/vitest-dev-vitest-q6njb9cq?file=src%2Findex.ts

```ts
import { format } from 'date-fns';
import { enUS } from 'date-fns/locale';

const date = new Date(2021, 2 - 1, 3, 12, 34, 56); // 2021-02-03 12:34:56

console.log(format(date, 'P', { locale: enUS }));
```

Even this simple script takes a long time to build - much longer to build than you would expect, just to import a single locale and the `format` function.

```
❯ npx vite build
vite v6.3.5 building for production...
✓ 829 modules transformed.
dist/index.html                 0.36 kB │ gzip: 0.26 kB
dist/assets/index-aRg9zBWn.js  21.52 kB │ gzip: 6.18 kB
✓ built in 1.21s

~/projects/vitest-dev-vitest-q6njb9cq 2s
```

829 modules - that sounds... high?

**Comment by mindplay-dk** [comment]
Why *does* this package have so many exports?

Tree shaking should handle one large package just fine, should it not?

**Comment by mindplay-dk** [comment]
I'm not sure I fully understand what the issue is, but I think it's that, when you import from the barrel file, the bundler goes to work bundling every single module in the package? because they're all reachable from there, and it tries to cache NPM packages?

If so, would it help if the published package had a bundled index file?

You could still have the individual submodules, of course - doesn't need to be a breaking change, just "ship everything twice". That sounds bad, but it probably won't increase the distributed package size much in practice - repeated stuff compresses very well. I think a lot of packages actually ship with two or more versions now, e.g. a module and a require.js script?

**Comment by mindplay-dk** [comment]
@caiotracera what are you using? a lot of the tools today are using the same underlying tools internally.

**Comment by caiotracera** [comment]
I also have a project using Vite 4.5.3 with `date-fns@3.6.0` that doesn't experience this performance issue.

**Comment by swierzbicki** [comment]
Any fixes or at least found the root of the issue? I upgraded to 3.x and also to 4.0 and both doubled my unit test execution time :( 

From 19min to 40min

**Comment by turenar** [comment]
## tl;dr

Importing a large number of small files is not performant. When running `vitest` without `--no-isolation`, all required files are **re-imported** every time a test file starts, which can result in significant overhead.

Possible mitigations:

* Import from `date-fns/funcName` instead of `date-fns`
* Bundle `date-fns` and set its path via `resolve.alias.date-fns` (`vite dev` does this automatically, but `vitest` appears **not** to)
    * Side effects: You can import only from `date-fns`

<details>
<summary>Scenarios and timing results</summary>

## Importing `date-fns`

```typescript
import {format} from "date-fns";
import {describe, expect, it} from "vitest";

describe('App', () => {
    it("format", () =>{
        expect(format(new Date(), "yyyy")).toEqual("2026");
    })
})
```

> Duration  1.24s (transform 30ms, setup 0ms, import 1.11s, tests 4ms, environment 0ms)

<img width="491" height="137" alt="Image" src="https://github.com/user-attachments/assets/34ea13e1-91a4-44f1-bd35-574d488326b9" />

Thie image above is a screenshot from `vitest --ui`.

## Importing `date-fns/format`

```typescript
import {format} from "date-fns/format";
import {describe, expect, it} from "vitest";

describe('App', () => {
    it("format", () =>{
        expect(format(new Date(), "yyyy")).toEqual("2026");
    })
})
```

> Duration  264ms (transform 29ms, setup 0ms, import 131ms, tests 3ms, environment 0ms)

<img width="491" height="137" alt="Image" src="https://github.com/user-attachments/assets/30ae1b4e-c5d9-4444-85ff-2af9b401f7ed" />

## Bundling and setting `resolve.alias.date-fns`

1. Bundle with `esbuild --bundle --format=esm --outfile=date-fns.mjs node_modules/date-fns/index.js`
2. Configure the alias in `vite.config.ts`

```typescript:vite.config.ts
export default defineConfig({
  // ...
  resolve: {
    alias: {
      "date-fns": "./path/to/date-fns.mjs"
    }
  }
});
```

```typescript
import {format} from "date-fns";
import {describe, expect, it} from "vitest";

describe('App', () => {
    it("format", () =>{
        expect(format(new Date(), "yyyy")).toEqual("2026");
    })
})
```

> Duration  287ms (transform 127ms, setup 0ms, import 161ms, tests 3ms, environment 0ms)

<img width="492" height="141" alt="Image" src="https://github.com/user-attachments/assets/c0274cf5-67a1-41ce-8e02-1c4d5ee0550f" />

## Notes

* `vite dev` bundles node_modules into several chunks at boot time. `vitest` doesn't.

</detail>

**Comment by mindplay-dk** [comment]
@turenar right, so the issue is the barrel file?

and I get that there are workarounds, but... bundlers are perfectly capable of tree shaking a bundled `index.js`, right? so is there any actual *advantage* to shipping a barrel file?

wouldn't it make more sense to ship a bundled `index.js` file like most projects do?

**Comment by turenar** [comment]
@mindplay-dk
I don't have a clear answer as to whether date-fns should be bundled.

For users who import from date-fns (without a subpath), bundling is unlikely to have any negative impact, as far as I know. Even now, bundlers without tree-shaking read and output everything as-is, since index.js imports all files.

My concern is about users who import from date-fns/funcName. date-fns currently allows subpath imports, but bundling may prevent imports from subpaths. In other words, bundling could break existing code in the wild.
(It might be fine if date-fns/a and date-fns/b pointed to the same chunk-12345.js, but I'm not an expert.)

(I accidentally posted this from my work account. Both is me.)

**Comment by mindplay-dk** [comment]
> It might be fine if date-fns/a and date-fns/b pointed to the same chunk-12345.js, but I'm not an expert

this most likely just leads to the same problem even for subpath imports.

I guess subpath imports are the only way to fix this.

I guess set up ESLint to ban imports from `date-fns` main module then?

ugh, what a PITA. 😌

**Comment by turenar** [comment]
I found [the prebundle option in Vitest](https://vitest.dev/config/deps). We don't have to bundle manually.

```typescript
{
  test: {
    deps: {
      optimizer: {
        ssr: { // or `client:`
          enabled: true,
          include: ["date-fns"]
} } } } }
```

This option seems to significantly improve test performance.

>   vite:deps Dependencies bundled in 93.19ms +94ms
>    Duration  192ms (transform 29ms, setup 0ms, import 48ms, tests 4ms, environment 0ms)

<img width="514" height="152" alt="Image" src="https://github.com/user-attachments/assets/f29ac527-3754-4fd8-81d6-5217dd4b3eef" />

One downside is that we shouldn’t mix `date-fns` and `date-fns/funcName` imports; otherwise, we won’t get the benefits of this optimization.

**Comment by mindplay-dk** [comment]
> One downside is that we shouldn’t mix `date-fns` and `date-fns/funcName` imports; otherwise, we won’t get the benefits of this optimization.

You can ban one or the other using a linter. But what happens when third-party packages use either one or the other?

I honestly think that offering this kind of "flexibility" (inconsistency) was a mistake.

Tree shaking is not some advanced or exotic feature, it's common  - it's available everywhere.

If this were my own package, I would remove the submodules, ship a major release, and be done with this.

This package is too important and too common to keep shipping something that creates this problem just about everywhere.

**Comment by fturmel** [comment]
Haven't been involved with this project for a while, but since this is still discussed and unresolved as of feb 2026 I wanted to surface PR #4023 as imo being the right solve and root cause analysis (CC @kossnocorp)

Wildcard re-exports means any bundler needs to scan most/every imports in the barrel file to locate the _actual_ source of a re-exported member. Probably not a big deal for a single pass bundle, but if Vite in dev mode or Vitest don't cache these resolutions there's probably tons of disk IO happening over and over again.

Reverting to v2's named export barrel file structure should be the way to go.
https://github.com/date-fns/date-fns/blob/v2.30.0/src/index.ts
https://github.com/date-fns/date-fns/blob/v3.0.0/src/index.ts <- problematic barrel file structure since v3

**Comment by mindplay-dk** [comment]
> Haven't been involved with this project for a while, but since this is still discussed and unresolved as of feb 2026 I wanted to surface PR [#4023](https://github.com/date-fns/date-fns/pull/4023) as imo being the right solve and root cause analysis

wow, yes, this makes a lot of sense! thank you for finding that. 🙂

> Reverting to v2's named export barrel file structure should be the way to go.

agreed. 👍

I'm not sure about the PR itself?

Just my point of view, and not that this PR is overwhelming at all, but I try not to put custom tooling in my own projects.

An ESLint rule should do the trick, I think? [example on ESLint playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXG4gIFwiZXJyb3JcIixcbiAge1xuICAgIFwic2VsZWN0b3JcIjogXCJFeHBvcnRBbGxEZWNsYXJhdGlvblwiLFxuICAgIFwibWVzc2FnZVwiOiBcIldpbGRjYXJkIHJlLWV4cG9ydHMgKGV4cG9ydCAqKSBhcmUgbm90IGFsbG93ZWQuXCJcbiAgfVxuXSAqL1xuXG5leHBvcnQgKiBmcm9tICcuL3doYXRldmVyJztcbmV4cG9ydCB7IGZvbyB9IGZyb20gJy4vc29tZXdoZXJlJztcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tcmVzdHJpY3RlZC1zeW50YXgiOlsiZXJyb3IiLHsic2VsZWN0b3IiOiJFeHBvcnRBbGxEZWNsYXJhdGlvbiIsIm1lc3NhZ2UiOiJXaWxkY2FyZCByZS1leHBvcnRzIChleHBvcnQgKikgYXJlIG5vdCBhbGxvd2VkLiJ9XX0sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==
)

also, that said, we don't know for certain that this actually fixes the issue - the author would need to publish a beta-release or something, and then we would need to test the package with various tooling.

I [posted a sandbox](https://github.com/date-fns/date-fns/issues/3806#issuecomment-3116693156) earlier that could be used for a quick preliminary test.

I did a quick test and installed version `2.30.0` in that project, and the `vitest` time dropped to less than half. 🙂

**Comment by turenar** [comment]
I completely agree with using named re-exports, since some tools (for example, [wyw-in-js](https://github.com/Anber/wyw-in-js)) can omit unused re-exports.
However, the current version of Vitest does not seem to eliminate them, possibly due to its design.

In addition, resolving `date-fns` to the **CommonJS** (CJS) build can improve performance even with `date-fns@^4`, according to [Vitest doc](https://main.vitest.dev/guide/profiling-test-performance.html#use-resolve-alias-to-redirect-imports). This suggests `date-fns@^2` is faster because that version provides only a CJS build.

But why are ESM imports slower?

Note:
[Someone recently contributed](https://github.com/vitest-dev/vitest/pull/9743) an update to [Vitest docs](https://main.vitest.dev/guide/profiling-test-performance.html#file-import), explicitly referencing date-fns by name. 😛

**Comment by mindplay-dk** [comment]
> But why are ESM imports slower?

I'm sure the addition of ESM is not the only difference between v2 and v4, right?

**Comment by turenar** [comment]
You're right. However, in my own experiments, I observed the same trend as described in the Vitest docs -- the difference between CJS and ESM appears to be significant.

* date-fns@**2.30.0 (CJS)**: 451ms
* date-fns@**4.1.0 (CJS)**: 296ms (even faster than v2)
* date-fns@**4.1.0 (ESM)**: 1140ms (significantly slow)

<details>
<summary>Test Conditions</summary>

`vitest/4.0.18 linux-x64 node-v22.22.0`

```typescript:vite.config.ts
// vite.config.ts
export default defineConfig({
  test: {
    fileParallelism: false,
  },
  resolve: {
    alias: [
      {
        find: /^date-fns$/,
        // replacement: join('./node_modules/.pnpm/date-fns@2.30.0/node_modules/date-fns/index.js'),
        // replacement: join('./node_modules/.pnpm/date-fns@4.1.0/node_modules/date-fns/index.cjs'),
        replacement: join('./node_modules/.pnpm/date-fns@4.1.0/node_modules/date-fns/index.js'),
      },
    ]
  },
})
```

```typescript
// test code
import {format} from "date-fns";
import {describe, expect, it} from "vitest";

describe('App', () => {
    it("format", () =>{
        expect(format(new Date(), "yyyy")).toEqual("2026");
    })
})
```

> `date-fns` -> `./node_modules/.pnpm/date-fns@2.30.0/node_modules/date-fns/index.js`
>    Duration  451ms (transform 32ms, setup 0ms, import 316ms, tests 3ms, environment 0ms)

> `date-fns` -> `./node_modules/.pnpm/date-fns@4.1.0/node_modules/date-fns/index.cjs`
>    Duration  296ms (transform 29ms, setup 0ms, import 157ms, tests 4ms, environment 0ms)

> `date-fns` -> `./node_modules/.pnpm/date-fns@4.1.0/node_modules/date-fns/index.js`
>    Duration  1.14s (transform 29ms, setup 0ms, import 989ms, tests 3ms, environment 0ms)

</details>

**Comment by mindplay-dk** [comment]
> the difference between CJS and ESM appears to be significant

Instead of resolving imports ahead of time (and possible omitting some of them) the CJS build does all of the "import work" at run-time rather than at build time.

The [ESM build](https://npmx.dev/package-code/date-fns/v/4.1.0/index.js):

```js
export * from "./add.js";
export * from "./addBusinessDays.js";
export * from "./addDays.js";
// ...
```

The compiler needs to use static analysis to make this work.

The [CJS build](https://npmx.dev/package-code/date-fns/v/4.1.0/index.cjs):

```js
var _index = require("./add.cjs");
Object.keys(_index).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (key in exports && exports[key] === _index[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index[key];
    },
  });
});
var _index2 = require("./addBusinessDays.cjs");
Object.keys(_index2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (key in exports && exports[key] === _index2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index2[key];
    },
  });
});
var _index3 = require("./addDays.cjs");
// ...
```

The run-time does the re-export work dynamically when the module is loaded.

So I'd expect a difference - I just would have never expected anything this drastic.

Anyhow, as your benchmark shows, I'd still expect `export { x, y, z } from 'foo'` to be slower at build-time than the CJS build, since even with named exports, it does need to parse all the files at build-time.

Named exports would clearly be an improvement, but it's definitely not "free" like CJS, where it doesn't need to do much more at build-time than just to verify that the file exists. 🤔
